core/matrix/diagonal.cpp

```diff
@@ -0,0 +1,269 @@
+
+    if (dynamic_cast<const Dense<ValueType> *>(b) &&
+        dynamic_cast<Dense<ValueType> *>(x)) {
+        exec->run(diagonal::make_right_apply_to_dense(
+            this, as<Dense<ValueType>>(b), as<Dense<ValueType>>(x)));
+    } else if (dynamic_cast<const Csr<ValueType, int32> *>(b) &&
+               dynamic_cast<Csr<ValueType, int32> *>(x)) {
+        exec->run(diagonal::make_right_apply_to_csr(
+            this, as<Csr<ValueType, int32>>(b), as<Csr<ValueType, int32>>(x)));
+    } else if (dynamic_cast<const Csr<ValueType, int64> *>(b) &&
+               dynamic_cast<Csr<ValueType, int64> *>(x)) {
+        exec->run(diagonal::make_right_apply_to_csr(
+            this, as<Csr<ValueType, int64>>(b), as<Csr<ValueType, int64>>(x)));
+    } else {
+        GKO_NOT_IMPLEMENTED;
+    }
+}
+
+
+template <typename ValueType>
+void Diagonal<ValueType>::apply_impl(const LinOp *alpha, const LinOp *b,
+                                     const LinOp *beta, LinOp *x) const
+{
+    if (dynamic_cast<const Dense<ValueType> *>(b) &&
+        dynamic_cast<Dense<ValueType> *>(x)) {
+        auto dense_x = as<Dense<ValueType>>(x);
+        dense_x->scale(beta);
+        dense_x->add_scaled(alpha, b);
+    } else {
+        GKO_NOT_IMPLEMENTED;
+    }
+}
+
+
+template <typename ValueType>
+std::unique_ptr<LinOp> Diagonal<ValueType>::transpose() const
+{
+    return this->clone();
+}
+
+
+template <typename ValueType>
+std::unique_ptr<LinOp> Diagonal<ValueType>::conj_transpose() const
+{
+    auto exec = this->get_executor();
+    auto tmp = Diagonal<ValueType>::create(exec, this->get_size()[0]);
+
+    exec->run(diagonal::make_conj_transpose(this, tmp.get()));
+    return std::move(tmp);
+}
+
+
+template <typename ValueType>
+void Diagonal<ValueType>::convert_to(Csr<ValueType, int32> *result) const
+{
+    auto exec = this->get_executor();
+    auto tmp = Csr<ValueType, int32>::create(
+        exec, this->get_size(), this->get_size()[0], result->get_strategy());
+    exec->run(diagonal::make_convert_to_csr(this, tmp.get()));
+    tmp->move_to(result);
+}
+
+
+template <typename ValueType>
+void Diagonal<ValueType>::move_to(Csr<ValueType, int32> *result)
+{
+    this->convert_to(result);
+}
+
+
+template <typename ValueType>
+void Diagonal<ValueType>::convert_to(Csr<ValueType, int64> *result) const
+{
+    auto exec = this->get_executor();
+    auto tmp = Csr<ValueType, int64>::create(
+        exec, this->get_size(), this->get_size()[0], result->get_strategy());
+    exec->run(diagonal::make_convert_to_csr(this, tmp.get()));
+    tmp->move_to(result);
+}
+
+
+template <typename ValueType>
+void Diagonal<ValueType>::move_to(Csr<ValueType, int64> *result)
+{
+    this->convert_to(result);
+}
+
+
+namespace {
+
+
+template <typename MatrixType, typename MatrixData>
+inline void read_impl(MatrixType *mtx, const MatrixData &data)
+{
+    // Diagonal matrices are assumed to be square.
+    GKO_ASSERT_EQ(data.size[0], data.size[1]);
```

Comment:
I think we have a [`GKO_ASSERT_IS_SQUARE_MATRIX`](https://github.com/ginkgo-project/ginkgo/blob/4df4dea1fbf142ddd3ee8aae9ff3f43c2a0829e1/core/factorization/ilu.cpp#L68) which throws `DimensionMismatch` which is better suited here. But maybe you have to add a `get_size()` method to `MatrixData`.